src/ristretto.rs

```diff
@@ -0,0 +1,257 @@
+// Copyright 2022 The Tari Project
+// SPDX-License-Identifier: BSD-3-Clause
+
+//! # RistrettoBulletProof
+//!
+//! Implementation of BulletProofs for the Ristretto group for Curve25519.
+
+use curve25519_dalek::{
+    constants::{RISTRETTO_BASEPOINT_COMPRESSED, RISTRETTO_BASEPOINT_POINT},
+    ristretto::{CompressedRistretto, RistrettoPoint},
+};
+
+use crate::{
+    generators::pedersen_gens::ExtensionDegree,
+    protocols::curve_point_protocol::CurvePointProtocol,
+    range_proof::RangeProof,
+    traits::{Compressable, Decompressable, FixedBytesRepr, FromUniformBytes},
+    PedersenGens,
+};
+
+/// A Bullet proof implentation using the Ristretto group.
+pub type RistrettoRangeProof = RangeProof<RistrettoPoint>;
+
+impl CurvePointProtocol for RistrettoPoint {}
+
+impl FixedBytesRepr for CompressedRistretto {
+    fn as_fixed_bytes(&self) -> &[u8; 32] {
+        CompressedRistretto::as_bytes(self)
+    }
+
+    fn from_fixed_bytes(bytes: [u8; 32]) -> Self {
+        Self(bytes)
+    }
+}
+
+impl Decompressable for CompressedRistretto {
+    type Decompressed = RistrettoPoint;
+
+    fn decompress(&self) -> Option<Self::Decompressed> {
+        CompressedRistretto::decompress(self)
+    }
+}
+
+impl FromUniformBytes for RistrettoPoint {
+    fn from_uniform_bytes(bytes: &[u8; 64]) -> Self {
+        RistrettoPoint::from_uniform_bytes(bytes)
+    }
+}
+
+impl Compressable for RistrettoPoint {
+    type Compressed = CompressedRistretto;
+
+    fn compress(&self) -> Self::Compressed {
+        RistrettoPoint::compress(self)
+    }
+}
+
+/// Generate `PedersenGens` with an `ExtensionDegree` of zero
+pub fn create_degree_zero_pedersen_gens() -> PedersenGens<RistrettoPoint> {
+    create_pedersen_gens_with_extension_degree(ExtensionDegree::Zero)
+}
+
```

Comment:
```suggestion
```

I think we can away with this interface as the intent is quite clear using the default function call.

Reply:
Would you rather rename this to `default_pedersen_gens`? Happy to do so.

The `Default` impl won't work on PedersenGens<P> because there is no generic idea of "default" for base points. If the idea of default PedersonGens were integral to bp+ we could create a new trait for this. I think there could well be further iterations on how we do generators.

Reply:
I actually meant having `create_pedersen_gens_with_extension_degree` is clear enough, or maybe as you suggest `default_pedersen_gens`, either will do, but my first choice would be just to remove that helper function.

Reply:
ah ok cool, yeah completely agree